slateratom/lib/avgpot.f90

```diff
@@ -0,0 +1,450 @@
+
+    allocate(rho(nRadial, nSpin), source=0.0_dp)
+
+    ! Build radial orbitals on Becke's Gauss-Chebyschev mesh
+    do iSpin = 1, nSpin
+      do ll = 0, max_l
+        nn = 0
+        do iDummy1 = 1, num_alpha(ll)
+          do iDummy2 = 1, poly_order(ll)
+            nn = nn + 1
+
+            do iRad = 1, nRadial
+              rad(iRad, nn, ll, iSpin) = wavefunction(cof(iSpin, ll, :, nn), alpha, num_alpha,&
+                  & poly_order, ll, abcissa(iRad))
+              radp(iRad, nn, ll, iSpin) = wavefunction_1st(cof(iSpin, ll, :, nn), alpha, num_alpha,&
+                  & poly_order, ll, abcissa(iRad))
+              radpp(iRad, nn, ll, iSpin) = wavefunction_2nd(cof(iSpin, ll, :, nn), alpha,&
+                  & num_alpha, poly_order, ll, abcissa(iRad))
+            end do
+
+            ! Electron density as occupation-weighted absolute squares of the orbitals
+            rho(:, iSpin) = rho(:, iSpin) + occ(iSpin, ll, nn) * rad(:, nn, ll, iSpin)**2
+
+          end do
+        end do
+      end do
+    end do
+
+    ! Check if established density integrates up to the correct number of electrons
+    electron_number = sum(weights * (rho(:, 1) + rho(:, 2)) * abcissa**2)
+    if (abs(sum(occ) - electron_number) > 1.0e-08_dp) then
+      call error("Average-potential: Mismatch in number of electrons.")
+    end if
+
+    ! Optionally check if the original kinetic energy is reproduced by the local quantities
+    if (present(kinetic_energy_ref)) then
+      kinetic_energy = getKineticEnergy(occ, abcissa, weights, max_l, num_alpha, poly_order, rad,&
+          & radp, radpp)
+      if (abs(kinetic_energy - kinetic_energy_ref) > 1.0e-06_dp) then
+        call error("Average-potential: Mismatch in kinetic energy.")
+      end if
+    end if
+
+    ! As an initial guess for the effective orbital energies we may just use the eigenvalues from
+    ! the converged GKS calculation:
+    eps(:, 0:, :) = eigval(:, 0:, :)
+    epsLast(:, 0:, :) = eps(:, 0:, :)
+
+    ! Start self-consistency iterations
+    lpSC: do iSC = 1, maxiter
+
+      ! Build average potential
+      avgPot(:,:) = 0.0_dp
+      do iSpin = 1, nSpin
+        do ll = 0, max_l
+          nn = 0
+          do iDummy1 = 1, num_alpha(ll)
+            do iDummy2 = 1, poly_order(ll)
+              nn = nn + 1
+
+              ! uses +laplace / 2
+              avgPot(:, iSpin) = avgPot(:, iSpin) + occ(iSpin, ll, nn)&
+                  & * (epsLast(iSpin, ll, nn) * rad(:, nn, ll, iSpin)**2&
+                  & + rad(:, nn, ll, iSpin) * (0.5_dp * radpp(:, nn, ll, iSpin)&
+                  & + radp(:, nn, ll, iSpin) / abcissa - 0.5_dp * ll * (ll + 1) / abcissa**2&
+                  & * rad(:, nn, ll, iSpin)))
+            end do
+          end do
+        end do
+      end do
+
+      ! Divide by the density, but try to avoid singular terms
+      do iSpin = 1, nSpin
+        where (rho(:, iSpin) > 0.0_dp)
+          avgPot(:, iSpin) = avgPot(:, iSpin) / rho(:, iSpin)
```

Comment:
Why not set avgPot to a constant (=0?) in an elsewhere clause for the case of rho being zero? Is the grid chosen so that the potential is only evaluated where the density is non-zero?

Reply:
`rho` can be/is zero at large distances, this statement is required in order to converge `avgPot` at all. I have added the elsewhere clause setting `avgPot` to zero where `.not. rho(:, iSpin) > 0.0_dp`.

Reply:
I think we discussed this last devel. meeting, therefore closing it.